File: android/app/src/main/java/org/findmyfam/services/LocalGroupAvatarStore.kt
package org.findmyfam.services

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ExifInterface
import android.net.Uri
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.io.File

/**
 * Local, per-user, per-device group avatars.
 *
 * Not shared with other members and not synced — purely a personal label on
 * your own view of a group. Stored as downscaled JPEGs in filesDir, keyed by
 * MLS group id. Call [init] from Application.onCreate() before use.
 *
 * (A future shared admin-controlled avatar can take precedence over this when it ships.)
 */
object LocalGroupAvatarStore {

    private const val MAX_DIMENSION = 256

    private lateinit var dir: File
    private val cache = HashMap<String, Bitmap>()

    private val _revision = MutableStateFlow(0)
    val revision: StateFlow<Int> = _revision.asStateFlow()

    fun init(context: Context) {
        dir = File(context.filesDir, "group_avatars").also { it.mkdirs() }
    }

    // MARK: - Read

    fun image(groupId: String): Bitmap? {
        cache[groupId]?.let { return it }
        val file = fileFor(groupId)
        if (!file.exists()) return null
        return BitmapFactory.decodeFile(file.absolutePath)?.also { cache[groupId] = it }
    }

    fun hasImage(groupId: String): Boolean =
        cache.containsKey(groupId) || fileFor(groupId).exists()

    // MARK: - Write

    fun setImage(context: Context, groupId: String, uri: Uri) {
        val rotation = context.contentResolver.openInputStream(uri)?.use { stream ->
            val exif = ExifInterface(stream)
            when (exif.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)) {
                ExifInterface.ORIENTATION_ROTATE_90 -> 90f
                ExifInterface.ORIENTATION_ROTATE_180 -> 180f
                ExifInterface.ORIENTATION_ROTATE_270 -> 270f
                else -> 0f
            }
        } ?: 0f

        val raw = context.contentResolver.openInputStream(uri)?.use {
            BitmapFactory.decodeStream(it)
        } ?: return

        val upright = if (rotation != 0f) {
            Bitmap.createBitmap(raw, 0, 0, raw.width, raw.height, Matrix().apply { postRotate(rotation) }, true)
        } else raw

        val scaled = downscale(upright)
        fileFor(groupId).outputStream().use { out ->
            scaled.compress(Bitmap.CompressFormat.JPEG, 80, out)
        }
        cache[groupId] = scaled
        _revision.value++
    }

    fun removeImage(groupId: String) {
        fileFor(groupId).delete()
        cache.remove(groupId)
        _revision.value++
    }

    fun removeAll() {
        dir.listFiles()?.forEach { it.delete() }
        cache.clear()
        _revision.value++
    }

    // MARK: - Private

    private fun fileFor(groupId: String) = File(dir, "$groupId.jpg")

    private fun downscale(src: Bitmap): Bitmap {
        val longest = maxOf(src.width, src.height)
        if (longest <= MAX_DIMENSION) return src
        val scale = MAX_DIMENSION.toFloat() / longest
        return Bitmap.createScaledBitmap(
            src,
            (src.width * scale).toInt(),
            (src.height * scale).toInt(),
            true
        )
    }
}

File: Sources/Services/LocalGroupAvatarStore.swift
import SwiftUI
import UIKit

/// Local, per-user, per-device group avatars.
///
/// These are **not** shared with other members and **not** synced across devices
/// — purely a personal touch on your own view of a group, so this needs no
/// protocol support, no encrypted blob distribution, and no admin control.
/// Stored as downscaled JPEGs in the app container, keyed by MLS group id.
///
/// (A future shared/admin-controlled avatar via `GroupDataUpdate.imageHash` can
/// take precedence over this local override when it ships.)
@MainActor
final class LocalGroupAvatarStore: ObservableObject {

    static let shared = LocalGroupAvatarStore()

    /// Bumped whenever an avatar changes so observing views re-read.
    @Published private(set) var revision = 0

    private var cache: [String: UIImage] = [:]
    private let dir: URL

    /// Longest-edge cap for stored thumbnails — keeps local storage tiny.
    private static let maxDimension: CGFloat = 256

    init() {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        dir = base.appendingPathComponent("GroupAvatars", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    }

    // MARK: - Read

    func image(for groupId: String) -> UIImage? {
        if let cached = cache[groupId] { return cached }
        guard let img = UIImage(contentsOfFile: fileURL(groupId).path) else { return nil }
        cache[groupId] = img
        return img
    }

    func hasImage(for groupId: String) -> Bool {
        cache[groupId] != nil || FileManager.default.fileExists(atPath: fileURL(groupId).path)
    }

    // MARK: - Mutate

    /// Downscale and save picked image data for a group.
    func setImage(data: Data, for groupId: String) {
        guard let raw = UIImage(data: data) else { return }
        let img = Self.downscaled(raw, maxDimension: Self.maxDimension)
        guard let jpeg = img.jpegData(compressionQuality: 0.8) else { return }
        try? jpeg.write(to: fileURL(groupId), options: .atomic)
        cache[groupId] = img
        revision += 1
    }

    func removeImage(for groupId: String) {
        try? FileManager.default.removeItem(at: fileURL(groupId))
        cache[groupId] = nil
        revision += 1
    }

    /// Clear all local avatars (e.g. on identity burn).
    func removeAll() {
        try? FileManager.default.removeItem(at: dir)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        cache.removeAll()
        revision += 1
    }

    // MARK: - Private

    private func fileURL(_ groupId: String) -> URL {
        // Group ids are hex, safe as a filename.
        dir.appendingPathComponent("\(groupId).jpg")
    }

    private static func downscaled(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
        let longest = max(image.size.width, image.size.height)
        guard longest > maxDimension else { return image }
        let scale = maxDimension / longest
        let newSize = CGSize(width: image.size.width * scale, height: image.size.height * scale)
        // Render at scale 1 so the stored JPEG is exactly `newSize` pixels, not
        // newSize × screen-scale (a @3x device would otherwise persist a 768px
        // thumbnail for a 256pt target).
        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1
        let renderer = UIGraphicsImageRenderer(size: newSize, format: format)
        return renderer.image { _ in image.draw(in: CGRect(origin: .zero, size: newSize)) }
    }
}
